src/librustc_save_analysis/lib.rs

```diff
@@ -324,17 +328,40 @@ impl<'l, 'tcx: 'l> SaveContext<'l, 'tcx> {
                     let sub_span = self.span_utils.sub_span_for_type_name(path.span);
                     filter!(self.span_utils, sub_span, typ.span, None);
 
+                    let kind = match *trait_ref {
+                        Some(_) => ImplKind::Direct,
+                        None => ImplKind::Inherent,
+                    };
+
+                    let impl_id = self.make_impl_id(&kind);
+                    let span = self.span_from_span(sub_span.unwrap());
+
                     let type_data = self.lookup_ref_id(typ.id);
                     type_data.map(|type_data| {
                         Data::RelationData(Relation {
                             kind: RelationKind::Impl,
-                            span: self.span_from_span(sub_span.unwrap()),
+                            span: span.clone(),
                             from: id_from_def_id(type_data),
                             to: trait_ref
                                 .as_ref()
                                 .and_then(|t| self.lookup_ref_id(t.ref_id))
                                 .map(id_from_def_id)
                                 .unwrap_or(null_id()),
+                            id: impl_id,
+                        },
+                        Impl {
+                            id: impl_id,
+                            kind: kind,
+                            span: span,
+                            value: String::new(),     // ???
+                            parent: None,             // ??? enclosing module?
+                            children: impls
+                                .iter()
+                                .map(|i| id_from_node_id(i.id, self))
+                                .collect(),
+                            docs: "".to_string(),     // ??? trait docs?
+                            sig: None,                // ??? trait sig?
+                            attributes: vec![],       // ??? trait attrs?
```

Comment:
There might be attributes on the impl that should be recorded here (I think). We don't need anything else. For docs and sig, we can leave them as they are (though I would prefer to use only one way to make an empty string).

Reply:
I fix

Reply:
@nrc Unless I'm looking in the wrong place, there doesn't seem to be any attrs on the `Impl` level (there are on the `ImplItem` level).

https://manishearth.github.io/rust-internals-docs/syntax/ast/enum.ItemKind.html